homeassistant/components/mastodon/config_flow.py

```diff
@@ -0,0 +1,167 @@
+"""Config flow for Mastodon."""
+
+from __future__ import annotations
+
+from typing import Any
+
+from mastodon.Mastodon import MastodonError
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, ConfigFlowResult
+from homeassistant.const import (
+    CONF_ACCESS_TOKEN,
+    CONF_CLIENT_ID,
+    CONF_CLIENT_SECRET,
+    CONF_NAME,
+)
+from homeassistant.helpers.selector import (
+    TextSelector,
+    TextSelectorConfig,
+    TextSelectorType,
+)
+
+from .const import CONF_BASE_URL, DEFAULT_NAME, DEFAULT_URL, DOMAIN, LOGGER
+from .utils import create_mastodon_instance
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(
+            CONF_NAME,
+            default=DEFAULT_NAME,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.TEXT)),
+        vol.Required(
+            CONF_BASE_URL,
+            default=DEFAULT_URL,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.URL)),
+        vol.Required(
+            CONF_CLIENT_ID,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.PASSWORD)),
+        vol.Required(
+            CONF_CLIENT_SECRET,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.PASSWORD)),
+        vol.Required(
+            CONF_ACCESS_TOKEN,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.PASSWORD)),
+    }
+)
+
+
+class MastodonConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+    config_entry: ConfigEntry
+
+    async def check_connection(
+        self,
+        base_url: str,
+        client_id: str,
+        client_secret: str,
+        access_token: str,
+    ) -> tuple[dict[str, str], dict[str, Any] | None]:
+        """Check connection to the Mastodon instance."""
+        try:
+            client = await self.hass.async_add_executor_job(
+                create_mastodon_instance,
+                base_url,
+                client_id,
+                client_secret,
+                access_token,
+            )
+            account = await self.hass.async_add_executor_job(
+                client.account_verify_credentials
+            )
```

Comment:
Rather just make this function not async and run this function in the executor loop